### PR TITLE
Add new option to save embedded relations changes in parent

### PR DESCRIPTION
**Pull request by JagdeepSingh**
This add a new option `embeds_many` to the `history_trackable_options` to save history of embedded objects in parent record itself.
e.g. if there are `User` and `Article` models as following:
```
class User
  include Mongoid::Document
  field :foo
  embeds_many :articles
  track_history embeds_many: [:articles]
end
```

```
class Article
  include Mongoid::Document
  field :bar
  embedded_in :user
end
```

This will save the `article.attributes` in addition to user's trackable fields to `user`'s history record whenever an user/article is created/updated/destroyed.

### CODE REVIEW

**Comment by JagdeepSingh** [comment]
Build failing with following error:
```
rbx-2.2.10 is not installed - installing.
Searching for binary rubies, this might take some time.
Requested binary installation but no rubies are available to download, consider skipping --binary flag.
Gemset '' does not exist, 'rvm rbx-2.2.10 do rvm gemset create ' first, or append '--create'.
```

**Comment by dblock** [comment]
This is great, but I think embedded objects are no different from other fields in some way, so I think the DSL should just be `on: [ :articles ]` and the code should be smart enough to know that those are embedded objects, relationships, etc. What do you think?

**Comment by JagdeepSingh** [comment]
@dblock `on:  [ :articles ]` seems as good as `embeds_many: [ :articles ]`. I will see if this can solve my requirement and do the change if possible.

**Comment by JagdeepSingh** [comment]
@dblock Now if a model has `track_history on: :all`, should it also track all embedded relations in addition to all the fields?

**Comment by JagdeepSingh** [comment]
If not, will the user have to specify all fields and embedded associations here?

**Comment by JagdeepSingh** [comment]
Also, if we track all fields and embedded relations in `:all`, then it will affect users who are already using it. They will be tracking the embedded objects in their applications, unintentionally.

**Comment by dblock** [comment]
I think we shouldn't touch `:all` for backward compatibility purposes. We can add `:all_fields` and `:all_relations` or similar in a separate PR if we want to.

**Comment by JagdeepSingh** [comment]
We can't touch `on: :all`, to support existing users. So, how would you like me to modify the DSL in this PR?. May be i can modify it to use a new option `associations: :all` (Other values can be `associations: :embeds_many`, `associations: :has_many`, `associations: :belongs_to`, etc and of course `[:articles, ...]`). Wdyt?

**Comment by dblock** [comment]
First, thanks for hanging on here. 

What I don't like about a DSL that you propose is that it's not naming "things", but "kinds of things". We can't touch `on: :all`, but we can do `on: [:all, :comments]`, right? It's not ideal because `:all` should, as you point out, be everything, but i think it's better than `associations: :comments` and much better than `associations: :has_many`.

**Comment by dblock** [comment]
We should alias `:all` to `:fields` and in the future add `:associations`.

**Comment by JagdeepSingh** [comment]
Ok. For now, i will make it `on: [:all, :comments, ...]`. And make `:all` an alias to `:fields`.

**Comment by johnnyshields** [comment]
@dblock the term `:relations` is better than `:associations` (in Mongoid parlance)

**Comment by dblock** [comment]
Thanks @JagdeepSingh, let us know when it's updated.

File: README.md [hunk 1]
```diff
@@ -156,6 +156,53 @@ Mongoid::History.disable do
...
+  include Mongoid::Timestamps
+  include Mongoid::History::Trackable
+
+  field           :title
+  field           :body
+  field           :rating
+  embeds_many     :comments
+
+  track_history   :on => [:title, :body],
+                  :embeds_many => [:comments],  # track comments as embedded attributes, default is []
```
Comment (dblock) [inline README.md]: This is still wrong.

**Comment by dblock** [comment]
LMK when this is ready to re-review, see comment above, fix the build, etc.

**Comment by JagdeepSingh** [comment]
This is still a WIP. I will let you know when to re-review it.

**Comment by JagdeepSingh** [comment]
@dblock i have done the functional changes from my end. Also fixed the build. You are free to review it now. I will be working on adding more rspecs till then.

**Comment by JagdeepSingh** [comment]
Will also add the remaining documentation comments on methods.

File: .travis.yml [hunk 2]
```diff
@@ -11,9 +11,11 @@ rvm:
   - 2.1.1
   - 2.0.0
   - 1.9.3
-  - rbx-2.2.10
   - jruby-19mode
 
+allow_failures:
+  - rvm: rbx-2
```
Comment (dblock) [inline .travis.yml]: You do need to re-add `rbx-2` though ;)

File: lib/mongoid/history/services/options_cleaner.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
+module Mongoid
+  module History
+    class OptionsCleaner
```
Comment (dblock) [inline lib/mongoid/history/services/options_cleaner.rb]: I think a better name could be just `Options`, no? We're not really cleaning anything, just dealing with options.

File: lib/mongoid/history/services/options_cleaner.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+          modifier_field: :modifier,
+          version_field: :version,
+          changes_method: :changes,
+          scope: scope,
+          track_create: false,
+          track_update: true,
+          track_destroy: false }
+      end
+
+      def clean(options = {})
```
Comment (dblock) [inline lib/mongoid/history/services/options_cleaner.rb]: Typically this is called `parse`.

File: spec/unit/services/options_cleaner_spec.rb [hunk 5]
```diff
@@ -0,0 +1,355 @@
+require 'spec_helper'
+
+describe Mongoid::History::OptionsCleaner do
+  class MyOptionsCleanerModel
```
Comment (dblock) [inline spec/unit/services/options_cleaner_spec.rb]: These classes are global, so you want to do

```ruby
let(:options_cleaner_model) do
   Class.new do
   end
end
```

Comment (dblock) [inline spec/unit/services/options_cleaner_spec.rb]: I do see that the rest of the code doesn't do that, so either change everything or just leave this alone, I am OK with it.

**Comment by dblock** [comment]
Thanks. I made a comment on class naming, I think we should rename it. 

There's still an entry in CHANGELOG needed.

File: lib/mongoid/history/services/options_cleaner.rb (hunk 4 above)
Comment (JagdeepSingh) [inline lib/mongoid/history/services/options_cleaner.rb]: Changing it to `parse`

File: lib/mongoid/history/services/options_cleaner.rb (hunk 3 above)
Comment (JagdeepSingh) [inline lib/mongoid/history/services/options_cleaner.rb]: Renaming it now

**Comment by JagdeepSingh** [comment]
@dblock Please check now.

**Comment by JagdeepSingh** [comment]
@dblock This is ready to be reviewed now.

**Comment by dblock** [comment]
Excellent work, thank you.

**Pull request merged**